spark/src/main/scala/geotrellis/spark/viewshed/IterativeViewshed.scala

```diff
@@ -0,0 +1,390 @@
+    * @param  epsilon      Rays within this many radians of horizontal (vertical) are considered to be horizontal (vertical)
+    * @param  touchedKeys  A set of SpatialKeys to keep track of which keys have been touched
+    */
+  def apply[K: (? => SpatialKey): ClassTag, V: (? => Tile)](
+    elevation: RDD[(K, V)] with Metadata[TileLayerMetadata[K]],
+    ps: Seq[Point6D],
+    maxDistance: Double,
+    curvature: Boolean = true,
+    operator: AggregationOperator = Or(),
+    epsilon: Double = (1/math.Pi),
+    touchedKeys: mutable.Set[SpatialKey] = null
+  )(implicit sc: SparkContext): RDD[(K, Tile)] with Metadata[TileLayerMetadata[K]] = {
+
+    ps.foreach({ p => require(p.length == 6) })
+
+    val md = elevation.metadata
+    val mt = md.mapTransform
+
+    val resolution = computeResolution(elevation)
+    logger.debug(s"Computed resolution: $resolution meters/pixel")
+
+    val bounds = md.bounds match {
+      case b: KeyBounds[K] => b
+      case _ => throw new Exception
+    }
+    val minKey: SpatialKey = bounds.minKey
+    val minKeyCol = minKey.col
+    val minKeyRow = minKey.row
+    val maxKey: SpatialKey = bounds.maxKey
+    val maxKeyCol = maxKey.col
+    val maxKeyRow = maxKey.row
+
+    val rays = new RayCatcher; sc.register(rays)
+
+    def validKey(key: SpatialKey): Boolean = {
+      ((minKeyCol <= key.col && key.col <= maxKeyCol) &&
+       (minKeyRow <= key.row && key.row <= maxKeyRow))
+    }
+
+    /**
+      * This function is used to create the `tileCallback` function
+      * that `R2Viewshed.compute` expects.  It receives packets of
+      * rays generated within the single-tile viewshed code and gives
+      * those to a Spark AccumulatorV2 so that they can be forwarded
+      * to interested neighbors.
+      */
+    def rayCatcherFn(key: SpatialKey, index: Int)(bundle: Bundle): Unit = {
+      val southKey = SpatialKey(key.col + 0, key.row + 1)
+      val westKey = SpatialKey(key.col + 1, key.row + 0)
+      val northKey = SpatialKey(key.col + 0, key.row - 1)
+      val eastKey = SpatialKey(key.col - 1, key.row + 0)
+
+      Map(
+        FromSouth() -> southKey,
+        FromWest() -> westKey,
+        FromNorth() -> northKey,
+        FromEast() -> eastKey
+      ).foreach({ case (dir, key) =>
+        if (validKey(key)) {
+          val rs = bundle.getOrElse(dir, throw new Exception)
+          if (rs.length > 0) {
+            val message = (key, index, dir, rs)
+            rays.add(message)
+          }
+        }
+      })
+    }
+
+    val info: Seq[PointInfo] = {
+      val fn = pointInfo(elevation)_
+      ps.zipWithIndex.map(fn)
+    }
+
+    val _pointsByKey: Map[SpatialKey, Seq[PointInfo]] =
+      info
+        .groupBy(_.key)
+        .toMap
+    val pointsByKey = sc.broadcast(_pointsByKey)
+    if (touchedKeys != null) touchedKeys ++= _pointsByKey.keys
+
+    val _pointsByIndex: Map[Int, PointInfo] =
+      info
+        .groupBy(_.index)
```

Comment:
Checking for my understanding, the `info` was created from `zipWithIndex` why not just take it in the order that its created?